sched/adaptdl_sched/allocator.py

```diff
@@ -38,11 +38,90 @@ class AdaptDLAllocator(object):
     def __init__(self, expander):
         self._core_api = kubernetes.client.CoreV1Api()
         self._objs_api = kubernetes.client.CustomObjectsApi()
+        self._custom_resource = ("adaptdl.petuum.com", "v1",
+                                 "", "adaptdljobs")
         self._cluster_expander = expander
         self._policy = PolluxPolicy()
+        # lock for the two corountines in run()
+        self.lock = asyncio.Lock()
 
     async def run(self):
+        # two functionality: (1) watch for new job and start if possible.
+        # (2) periodically optimize existing jobs
+        await asyncio.gather(
+            self._new_job_allocate(),
+            self._periodic_optimize()
+        )
+
+    async def _new_job_allocate(self):
+        async with kubernetes.watch.Watch() as watch:
+            while True:
+                async for event in watch.stream(
+                        self._objs_api.list_namespaced_custom_object,
+                        *self._custom_resource, timeout_seconds=60):
+                    # if there is an arriving job
+                    if event["type"] == "ADDED":
+                        await self.lock.acquire()
+                        # re-read the job , compared with the previous readjob
+                        job_old = event["object"]
+                        namespace_old = job_old["metadata"]["namespace"]
+                        name_old = job_old["metadata"]["name"]
+
+                        job_list = await \
+                            self._objs_api.list_namespaced_custom_object(
+                                "adaptdl.petuum.com", "v1", "", "adaptdljobs")
+
+                        job = None
+                        for job_new in job_list["items"]:
+                            namespace_new = job_new["metadata"]["namespace"]
+                            name_new = job_new["metadata"]["name"]
+                            # found the old job
+                            if namespace_new == namespace_old \
+                                    and name_new == name_old:
+                                job = job_new
+                                break
+                        else:
+                            # This job is don
+                            self.lock.release()
+                            continue
+
+                        # some other coroutine has handled this
+                        LOG.info(job.get("status", {}).get("allocation", []))
+                        if len(job.get("status", {})
+                                .get("allocation", [])) != 0:
```

Comment:
Just `not job.get("status", {}).get("allocations")` is enough.

Also I think we need to add a `job.get("status", {}).get("group") is None` because the current condition doesn't catch the case where a job is allocated some resources, de-allocated all resources, and allocated again.